File: LocalWSLUbuntu_May2021/cs444/Barbershop.py
import sys
import threading
import time 
import random

n = 0
customers = 0
closing = 0

mutex = threading.Lock()
customerSem = threading.Lock()
barberSem = threading.Lock()
customerDone = threading.Lock()
barberDone = threading.Lock()

def kickOutCustomers():
    for i in customerThreads:
        balk(customerThreads[i])

def cutHair():
    print("Barber is cutting hair...\n")

def getHairCut(self):
    print("Thread " + str(self) + " is getting its hair cut...\n")

def balk(self):
    print("Screw you guys, Thread " + str(self) + " is going home\n")
    mutex.release()
    self.kill()

def barberRunner():
    print("Barber arrived\n")
    start = time.thread_time()
    shift = time.thread_time() - time.thread_time()

    while (shift < 15.0):
        customerSem.acquire()
        barberSem.release()

        cutHair()
        
        customerDone.acquire()
        barberDone.release()

        shift = time.thread_time() - time.thread_time()

    kickOutCustomers()
    closing = 1
    print("Barber is leaving!\n")
    barber.kill()

def customerRunner():
    self = threading.current_thread()

    mutex.acquire()
    if customers >= n:
        balk(self)
    customers += 1
    mutex.release()

    customerSem.release()
    barberSem.acquire()

    getHairCut(self)

    customerDone.release()
    barberDone.acquire()

    mutex.acquire()
    customers -= 1
    mutex.release()

    self.kill()

barber = threading.Thread(target=barberRunner)
customerThreads = []

if __name__ == "__main__":

    if len(sys.argv) != 2:
        sys.exit("./BarbershopP n\nWhere n = number of seats\n")

    n = sys.argv[1]

    barberSem.acquire()
    barberDone.acquire()
    customerSem.acquire()
    customerDone.acquire()

    barber.start()

    while not (closing):
        random.seed(time.time())
        x = random.randint(0,6)

        print("Creating " + str(x) + " threads...\n")

        time.sleep(3)

        for i in range (1, x):
            t = threading.Thread(target=customerRunner)
            customerThreads.append(t)
            t.start()

    barber.join()


    

File: LocalWSLUbuntu_May2021/cs444/Barbershop.c
#include "stdio.h"
#include "stdlib.h"
#include "unistd.h"
#include "pthread.h"
#include "semaphore.h"

int n;
int customers;
int closing;

sem_t mutex;
sem_t customerSem;
sem_t barberSem;
sem_t customerDone;
sem_t barberDone;

pthread_t barber;
pthread_t *customer;

void shutdown() {
    pthread_exit(&barber);
    for(int i = 0; i > 7; i++) {
        pthread_exit(&customer[i]);
    }

    sem_destroy(&mutex);
    sem_destroy(&customerSem);
    sem_destroy(&barberSem);
    sem_destroy(&customerDone);
    sem_destroy(&barberDone);
}

void cutHair() {
    printf("Barber is cutting hair\n");
}

void getHairCut(pthread_t name) {
    printf("Thread %ld is getting it's hair cut\n", name);
}

void balk(pthread_t name) {
    printf("Screw you guys, Thread %ld is goin' home!\n", name);
    sem_post(&mutex);
    pthread_exit(&name);
}

void * barberRunner(void * args) {
    while(!closing) {
            
        sem_wait(&customerSem);
        sem_post(&barberSem);

        cutHair();

        sem_wait(&customerDone);
        sem_post(&barberDone);
    }

    printf("Barber is leaving\n");
    pthread_t self = pthread_self();
    pthread_exit(&self);
}

void * customerRunner(void * args) {
    pthread_t self = pthread_self();

    sem_wait(&mutex);
    if(customers >= n) 
        balk(self);
    customers++;
    sem_post(&mutex);

    sem_post(&customerSem);
    sem_wait(&barberSem);

    getHairCut(self);

    sem_post(&customerDone);
    sem_wait(&barberDone);

    sem_wait(&mutex);
    customers--;
    sem_post(&mutex);

    pthread_exit(&self);
}

int main(int argc, char *args[]) {

    if(argc != 2) {
        printf("./BarbershopC n\nWhere n = number of seats\n");
        exit(0);
    }
    char *input = args[1];
    n = atoi(input);
    customers = 0;
    closing = 0;

    sem_init(&mutex, 0, 1);
    sem_init(&customerSem, 0, 1);
    sem_init(&barberSem, 0, 1);
    sem_init(&customerDone, 0, 1);
    sem_init(&barberDone, 0, 1);

    barber;
    customer = malloc(7 * sizeof(pthread_t));
    
    if(pthread_create(&barber, NULL, barberRunner, NULL) != 0) {
        printf("Error creating barber thread\n");
        shutdown();
        exit(0);
    }

    sem_post(&mutex);

    int timesRun = 5;
    for (int i = 0; i < timesRun; i++) {
        sleep(3);


        time_t t;
        srand((unsigned) time(&t));
        int x = (rand() % 5) + 1;

        printf("%d new threads have arrived!\n", x);

        if(i == timesRun - 1) {
           closing = 1;
        }

        for(int i = 0; i < x; i++) {
            if(pthread_create(&customer[i], NULL, customerRunner, NULL) != 0) {
                printf("Error creating customer thread\n");
                shutdown();
                exit(0);
            } 
        }
    }

    pthread_join(barber, NULL);
    shutdown();
    free(customer);
    return 0;
}